### PR TITLE
indexer: polish, step 0

### PR DESCRIPTION
**Pull request by hosted-fornet**
220815

A variety of changes were packed into this PR. Here is a summary:

* Add infrastructure and querying of only most recent batch. We expect users to frequently just want to know the current state of something -- and we also need to be able to conveniently serve updates to subscribers.
    * Added `newest-batch-by-town` to state, which is overwritten by each new batch that comes in.
    * Added endpoints to query only most recent block as `/newest/grain/...` as opposed to `/grain/...` which will return the entire history of that grain.
    * Made use of `newest-batch-by-town` to greatly simplify subscriber card creation logic.
* Got rid of `lib/indexer-bowl.hoon` which was a more-or-less hacky means of supplying rice-specific data to JSON scries that was no longer serving a purpose following our change to use `lump`s.
* Made catching up a new indexer to the current state on an old indexer somewhat more robust and a lot more correct.

The commits are reasonably atomic, and since there are changes with a number of purposes here, it might make sense to review commit-by-commit.

I tested as follows:

```
::  on ~zod
:rollup|activate
:sequencer|init our 0x0 0xc9f8.722e.78ae.2e83.0dd9.e8b9.db20.f36a.1bc4.c704.4758.6825.c463.1ab6.daee.e608
:indexer &set-sequencer [our %sequencer]
:indexer &set-rollup [our %rollup]
:uqbar|set-sources 0x0 our
:wallet &zig-wallet-poke [%import-seed 'uphold apology rubber cash parade wonder shuffle blast delay differ help priority bleak ugly fragile flip surge shield shed mistake matrix hold foam shove' 'squid' 'nickname']
:wallet &zig-wallet-poke [%submit from=0x7a9a.97e0.ca10.8e1e.273f.0000.8dca.2b04.fc15.9f70 to=0x74.6361.7274.6e6f.632d.7367.697a town=0x0 gas=[1 1.000.000] [%give to=0xd6dc.c8ff.7ec5.4416.6d4e.b701.d1a6.8e97.b464.76de amount=123.456 grain=0x89a0.89d8.dddf.d13a.418c.0d93.d4b4.e7c7.637a.d56c.96c0.7f91.3a14.8174.c7a7.71e6]]
:sequencer|batch

:: on ~nec
-zig!indexer-catchup 0x0 ~zod ~zod ~zod
:: OR
:indexer &indexer-catchup [~zod %indexer]
:indexer &set-sequencer [~zod %sequencer]
:indexer &set-rollup [~zod %rollup]
:uqbar|set-sources our 0x0 our
```

And then querying to make sure the indexer state is the same.

### CODE REVIEW

File: app/indexer.hoon [hunk 1]
```diff
@@ -335,6 +336,23 @@
       =/  hash=@ux  (slav %ux i.t.args)
       =/  =update:ui  (serve-update query-type hash)
       (make-peek-update update)
+    ::
+        $?  [%x %newest-egg @ ~]
```
Comment (tacryt-socryp) [inline app/indexer.hoon:340]: never seen this syntax used before, but it's beautiful

**Comment by hosted-fornet** [comment]
Ready for review. Not sure who wants to review, but feel free to add self. The main thing that I want to do to indexer that is not in this PR is add `grain-eggs` back in, but I think that requires `+lump-to-mold` since I need to be able to cast the `yolk` noun in order to read out the `grain` IDs.

@tacryt-socryp @dr-frmr

**Comment by dr-frmr** [comment]
Reviewing, looking good.. @hosted-fornet will the wallet need any changes to only get latest grain version? right now it's using the indexer's `update` type, have you tested to see if this still works?

**Comment by hosted-fornet** [comment]
Good point, let me take a look.

**Comment by hosted-fornet** [comment]
Repeating some discussion in DMs here to put on the record: I'm going to make a new branch pointing here to work on %wallet and %uqbar. The behavior here of %wallet and %uqbar is functional, but not optimal: e.g. if you query %uqbar for a grain, you get the entire grain history rather than the newest.

**Review by dr-frmr — APPROVED** [review]
Can merge now or wait and merge other work here

**Pull request merged**